### PR TITLE
Operator: Update MON creation to use node hostnames

### PR DESCRIPTION
**Pull request by dgonzalezruiz**
This allows the correct functionality of the cluster in kubernetes setups where the node name differs from the actual hostname.

This PR implements the solution to #1264

### CODE REVIEW

File: pkg/operator/mon/health.go [hunk 1]
```diff
@@ -161,7 +161,7 @@ func (c *Cluster) checkMonsOnSameNode() (bool, error) {
 	nodesUsed := map[string]struct{}{}
 	for name, node := range c.mapping.Node {
 		// when the node is already in the list we have more than one mon on that node
-		if _, ok := nodesUsed[node.Name]; ok {
+		if _, ok := nodesUsed[node.Address]; ok {
```
Comment (travisn) [inline pkg/operator/mon/health.go]: Instead of the `Name`, we will need to use the Label that has the name `kubernetes.io/hostname`. If we don't find that label then we could fall back to the name. So wherever we need the node name we would probably want to call a helper method that does this.

Comment (dgonzalezruiz) [reply pkg/operator/mon/health.go]: Fully agree! Something that receives the `node` objects and returns `Address` if defined, and `Name` otherwise.

**Comment by dgonzalezruiz** [comment]
The latest state passes integration tests in my machine, but does not pass unit tests either. However, it seems to be unrelated to my changes. Any suggestion @travisn ?

**Comment by jbw976** [comment]
@dgonzalezruiz the unit test failure looks related to the addition of the new hostName property:
```
--- FAIL: TestSaveMonEndpoints (0.00s)
	assertions.go:237: 
                        
	Error Trace:	mon_test.go:247
		
	Error:		Not equal: 
			expected: "{\"node\":{\"mon1\":{\"Name\":\"node0\",\"Address\":\"1.1.1.1\"}},\"port\":{\"node0\":12345}}"
			received: "{\"node\":{\"mon1\":{\"Name\":\"node0\",\"Address\":\"1.1.1.1\",\"Hostname\":\"\"}},\"port\":{\"node0\":12345}}"
```
The unit test would also need to add that property to it's expected output that it's testing against.

File: pkg/operator/mon/util.go [hunk 2]
```diff
@@ -302,3 +302,13 @@ func extractKey(contents string) (string, error) {
 	}
 	return secret, nil
 }
+
+func getNodeId(node *NodeInfo) (string, error) {
```
Comment (jbw976) [inline pkg/operator/mon/util.go]: this function doesn't look like it can return an error at all, so the return type should probably just be string

File: pkg/operator/mon/spec.go [hunk 3]
```diff
@@ -91,7 +91,7 @@ func (c *Cluster) makeMonPod(config *monConfig, nodeName string) *v1.Pod {
 	podSpec := v1.PodSpec{
 		Containers:    []v1.Container{container},
 		RestartPolicy: v1.RestartPolicyAlways,
-		NodeSelector:  map[string]string{apis.LabelHostname: nodeName},
+		NodeSelector:  map[string]string{apis.LabelHostname: nodeId},
```
Comment (jbw976) [inline pkg/operator/mon/spec.go]: if nodeID is a name instead of a hostname, then we may need to use different key here instead of `LabelHostName`.

File: pkg/operator/mon/mon.go [hunk 4]
```diff
@@ -328,18 +329,22 @@ func getNodeInfoFromNode(n v1.Node) (*NodeInfo, error) {
 			break
 		}
 	}
+	nr.Hostname = n.Labels["kubernetes.io/hostname"]
 	if nr.Address == "" {
-		return nil, fmt.Errorf("no IP given for node %s", nr.Name)
+		return nil, fmt.Errorf("no IP given for node %s", nr.Address)
```
Comment (jbw976) [inline pkg/operator/mon/mon.go]: inside this block, `nr.Address` will be an empty string, so using that in the log message won't be as helpful as using a different (non empty) field on the `NodeInfo` struct.

File: pkg/operator/mon/mon.go [hunk 5]
```diff
@@ -328,18 +329,22 @@ func getNodeInfoFromNode(n v1.Node) (*NodeInfo, error) {
 			break
 		}
 	}
+	nr.Hostname = n.Labels["kubernetes.io/hostname"]
```
Comment (jbw976) [inline pkg/operator/mon/mon.go]: there is a const value you can reference for this instead of using a hard coded string: `apis.LabelHostname`

File: pkg/operator/mon/mon_test.go [hunk 6]
```diff
@@ -234,8 +234,9 @@ func TestSaveMonEndpoints(t *testing.T) {
 	c.clusterInfo.Monitors["mon1"].Endpoint = "2.3.4.5:6790"
 	c.maxMonID = 2
 	c.mapping.Node["mon1"] = &NodeInfo{
-		Name:    "node0",
-		Address: "1.1.1.1",
+		Name:     "node0",
+		Address:  "1.1.1.1",
+                Hostname: "",
```
Comment (jbw976) [inline pkg/operator/mon/mon_test.go]: style nit: this line looks indented more than the others.

File: pkg/operator/mon/spec.go (hunk 3 above)
Comment (dgonzalezruiz) [reply pkg/operator/mon/spec.go]: If name is used as a node identifier then there is no way I can think of to make a node discrimination in the mon replicasets. Nodeselectors can only be labels https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#interlude-built-in-node-labels 

What do you think that would be the best solution, then? Delete the nodeselector?

**Comment by dgonzalezruiz** [comment]
Thanks a ton for your feedback! @jbw976 
I will update the MR after you answer the question above

File: pkg/operator/mon/spec.go (hunk 3 above)
Comment (charleslieferando) [reply pkg/operator/mon/spec.go]: @jbw976 ? :)

File: pkg/operator/mon/mon_test.go [hunk 7]
```diff
@@ -234,8 +234,9 @@ func TestSaveMonEndpoints(t *testing.T) {
 	c.clusterInfo.Monitors["mon1"].Endpoint = "2.3.4.5:6790"
 	c.maxMonID = 2
 	c.mapping.Node["mon1"] = &NodeInfo{
-		Name:    "node0",
-		Address: "1.1.1.1",
+		Name:     "node0",
+		Address:  "1.1.1.1",
+		Hostname: "",
```
Comment (kokhang) [inline pkg/operator/mon/mon_test.go]: can we add a test where the Hostname is not empty? That way we can test this path.

File: pkg/operator/mon/mon.go [hunk 8]
```diff
@@ -337,9 +339,9 @@ func getNodeInfoFromNode(n v1.Node) (*NodeInfo, error) {
 func (c *Cluster) startPods(mons []*monConfig) error {
 	for _, m := range mons {
 		node, _ := c.mapping.Node[m.Name]
-
+		nodeId := getNodeId(node)
```
Comment (kokhang) [inline pkg/operator/mon/mon.go]: There are more places in this file where `node.Name` is being referenced. Did you ensure that those also work when the Hostname differs from the Name? Im starting to think this is a much bigger undertaking that we previously thought.

File: pkg/operator/mon/mon_test.go (hunk 7 above)
Comment (dgonzalezruiz) [reply pkg/operator/mon/mon_test.go]: Sure! WIll act accordingly. Thx for the feedback! :)

File: pkg/operator/mon/mon.go (hunk 8 above)
Comment (dgonzalezruiz) [reply pkg/operator/mon/mon.go]: Actually I think it's even simpler. The Hostname of a kubernetes cluster is populated by kubernetes and therefore it is the only one we should use for node selection. As explained in https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetesiohostname , we can assume its existence as a fact.

**Comment by travisn** [comment]
@dgonzalezruiz Any chance you are able to look at this again soon? It would be great to get this in. Thanks!

File: pkg/operator/mon/spec.go (hunk 3 above)
Comment (PaulCapestany) [reply pkg/operator/mon/spec.go]: Curious to hear if @jbw976 has any opinions on @dgonzalezruiz's question. I've also ran into issue #1264 and it unfortunately has seemed to be a total blocker for getting rook up and running properly..

**Comment by travisn** [comment]
This PR is replaced by #1409

**Comment by PaulCapestany** [comment]
@jbw976 for others who might be interested in potentially contributing to rook, it might be useful to hear what might have gone right/wrong with this PR (from rook core contributors’ perspectives). 

Being new to rook (and Ceph) myself, but, being very interested in its potential, I’d like to have a better sense of the rules of engagement, where the bar is at, etc :)

**Comment by travisn** [comment]
Yes @dgonzalezruiz i want to thank you for getting this going and providing the basis for the change. I should have provided more background before closing it. The motivation was simply that we haven't heard from you in a while and wanted to unblock some members of the community who were hitting this. This morning we discussed in the biweekly community meeting as well. If you see this message soon, you're welcome to reopen your PR and we can finish it instead of merging #1409. 

@PaulCapestany great question on the community engagement. We value community contributions a great deal and I hope that moving forward on this item doesn't give the impression otherwise. Our code of conduct can be found [here](https://github.com/rook/rook/blob/master/CODE_OF_CONDUCT.md). Looking forward to your feedback and contributions as well!

**Comment by travisn** [comment]
Ok, got confirmation that @dgonzalezruiz will work on this today. Thanks!

File: pkg/operator/mon/spec.go (hunk 3 above)
Comment (jbw976) [reply pkg/operator/mon/spec.go]: From all the environments I've seen, the `kubernetes.io/hostname` label has always been set on the nodes, regardless of if it matches the node name field or not.  Your environment has this label set also.  I suspect that since the `getNodeId` function has a preference for the hostname field, which comes from the `kubernetes.io/hostname` label, that this selector will work as intended.

File: pkg/operator/mon/mon.go [hunk 9]
```diff
@@ -328,6 +329,7 @@ func getNodeInfoFromNode(n v1.Node) (*NodeInfo, error) {
 			break
 		}
 	}
+	nr.Hostname = n.Labels[apis.LabelHostname]
```
Comment (jbw976) [inline pkg/operator/mon/mon.go]: could you move this line up to just after 324 where the `nr.Name` field is being set?  It's right in the middle of the `nr.Address` setting/checking, which slightly disrupts the readability.

**Comment by dgonzalezruiz** [comment]
I accidentally borked this PR; #1409 will be the way to go.
